src/plugins/data/public/antlr/dql/code_completion.ts

```diff
@@ -174,21 +142,123 @@ export const getSuggestions = async ({
     // gets candidates at specified token index
     const candidates = core.collectCandidates(cursorIndex);
 
-    const completions = [];
+    // manually remove NOT from candidates when cursor is in a phrase
+    if (tokenStream.get(cursorIndex).type === DQLParser.PHRASE) {
+      candidates.tokens.delete(DQLParser.NOT);
+    }
+
+    const completions: QuerySuggestion[] = [];
 
     // check to see if field rule is a candidate. if so, suggest field names
     if (candidates.rules.has(DQLParser.RULE_field)) {
       completions.push(...findFieldSuggestions(indexPattern));
     }
 
+    interface FoundLastValue {
+      field: string | undefined;
+      value: string | undefined;
+    }
+
+    // visitor for parsing the current query
```

Comment:
The purpose of this 'visitor' here is to traverse the AbstractSyntaxTree generated by DQLParser and return what it thinks is the associated `field` and associated `value`, if there is one that is partially complete. 

`defaultResult` and `aggregateResult` define the default behavior at each node of what to do with the return values from visiting a particular node's children.

The other methods, in the format `visit[RULE_NAME]` override the default rule visitors in order to define custom behavior.

The starting point of the custom rule visitor functions is `KeyValueExpression`, which will match the parts of the dql query that have the format `field: value`, which is the only instance in which we want to provide value suggestions. The value here can be as simple as a 'tokenSearch' or a 'phrase', from which we can easily find the current value if there is one. However, it can also be a `groupExpression`, for which we have to define more custom behavior on those rule visitors. The general strategy is to enter and return nodes that contain the cursor position. 

Ultimately, the found field + value combinations are returned up the tree, giving deference to combinations found on the right side since more recent query input is on that side.